### PR TITLE
feat: Implement `Serialize` trait for each `Node`

### PR DESCRIPTION
**Pull request by magurotuna**
This patch implements `Serialize` trait for each `Node`. This will be super useful for making deno_lint's plugin system I'm working on more practical.

The difference may be a bit hard to follow, but how it works can be seen in test cases. See `rs-lib/tests/test.rs` and ` rs-lib/tests/expected/*.json`. 

For now I don't need to have `Deserialize` trait implemented, so I didn't do that.

### CODE REVIEW

**Comment by dsherret** [comment]
@magurotuna it looks good, but isn't this already handled by swc? Couldn't the `.inner` be serialized?

File: rs-lib/src/generated.rs [hunk 1]
```diff
@@ -6397,23 +6448,20 @@ fn get_view_for_array_pat<'a>(inner: &'a swc_ast::ArrayPat, parent: Node<'a>, bu
...
-impl<'a> ArrowExpr<'a> {
-  pub fn is_async(&self) -> bool {
-    self.inner.is_async
-  }
-
-  pub fn is_generator(&self) -> bool {
-    self.inner.is_generator
-  }
+  pub is_async: bool,
+  pub is_generator: bool,
```
Comment (dsherret) [inline rs-lib/src/generated.rs:6464]: Previously the methods were used to prevent storing information twice. I'd rather this be micro-optimized and use as little memory as possible, but I don't know what impact this has in practice.

**Comment by dsherret** [comment]
I think perhaps we could still implement Serialize, but it would be turned on via a cargo feature and would serialize the `inner` property with swc. Thoughts? (I'm not sure about needing to make this a cargo feature, but I don't need it in dprint-plugin-typescript)

**Comment by magurotuna** [comment]
@dsherret 

> @magurotuna it looks good, but isn't this already handled by swc? Couldn't the .inner be serialized?

Yeah it is, but for the following reason I'd be happy if the nodes of `dprint-swc-ecma-ast-view` implemented `Serialize`.

The reason is, swc's implementation for `Serialize` is a little bit tricky. I mean, if we try to serialize the following code:

```ts
function foo(a) { }
```

then we'll get the result as follows via swc's `Serialize`:

```json
{
  "type": "FunctionDeclaration",
  "identifier": {
    "type": "Identifier",
    "value": "foo",
    "optional": false
  },
  "declare": false,
  "params": [
    {
      "type": "Parameter",
      "decorators": [],
      "pat": {
        "type": "Identifier",
        "value": "a",
        "optional": false,
        "typeAnnotation": null
      }
    }
  ],
  "decorators": [],
  "body": {
    "type": "BlockStatement",
    "stmts": []
  },
  "generator": false,
  "async": false,
  "typeParameters": null,
  "returnType": null
}
```

Note: I omitted `span`s for readability.

On the other hand, if we use `dprint-swc-ecma-ast-view`'s `Serialize`, we'll get:

```json
{
  "nodeKind": "FnDecl",
  "ident": {
    "nodeKind": "Ident",
    "sym": "foo",
    "optional": false
  },
  "function": {
    "nodeKind": "Function",
    "params": [
      {
        "nodeKind": "Param",
        "decorators": [],
        "pat": {
          "nodeKind": "BindingIdent",
          "id": {
            "nodeKind": "Ident",
            "sym": "a",
            "optional": false
          },
          "typeAnn": null
        }
      }
    ],
    "decorators": [],
    "body": {
      "nodeKind": "BlockStmt",
      "stmts": []
    },
    "typeParams": null,
    "returnType": null,
    "isGenerator": false,
    "isAsync": false
  },
  "declare": false
}
```

There are a couple of differences, but the point is, the former _doesn't_ have `function` field, while the latter _does_ have.
This is because swc's `Serialize` implementation [uses #[serde(flatten)] here](https://docs.rs/swc_ecma_ast/0.40.0/src/swc_ecma_ast/decl.rs.html#45). Thus, it can be said that the serialization result of swc doesn't exactly reflect its original structure. This fact can become a problem when we traverse AST nodes using `dprint-swc-ecma-ast-view`, because if we attempt to call `.children()` on the above `FnDecl`, we get a list of `Ident` and `Function`, but the serialization result via swc doesn't have `Function`.

In deno_lint's plugin system, users will be able to create their own plugins in JavaScript or TypeScript, with serialized AST information passed from Rust side. And the capability of upward and downward traversing on AST will still be provided by `dprint-swc-ecma-ast-view` through Deno's Op. Therefore, I want the serialization result of AST to match with its corresponding struct's shape in Rust as much as possible.

> I think perhaps we could still implement Serialize, but it would be turned on via a cargo feature and would serialize the inner property with swc. Thoughts? (I'm not sure about needing to make this a cargo feature as perhaps it's not a big deal to just always include it, but I don't need it in dprint-plugin-typescript)

Making this a cargo feature sounds good to me :) But unfortunately, as I mentioned above, using inner property's `Serialize` doesn't satisfy my needs.

File: rs-lib/src/generated.rs (hunk 1 above)
Comment (magurotuna) [reply rs-lib/src/generated.rs:6464]: I'm not sure either... 
It seems like most values that are turned into fields from methods are lightweight types like bool, so I hope it's not a big deal.

Comment (dsherret) [reply rs-lib/src/generated.rs:6464]: A few extra fields could bump up the size of the struct and when there are many nodes that could increase the memory usage a decent amount (it is a micro-optimization, but again, I'd like this to be micro-optimized). I'm leaning to not changing this for the time being as we could implement serialization without it though, I wouldn't start on it yet... still want to think understand/think about this.

Comment (magurotuna) [reply rs-lib/src/generated.rs:6464]: So it is acceptable to implement `Serialize` but unacceptable to have extra fields, right?
I just came up with an idea to work around it - for instance, we use `#[serde(into = "FnDeclForSerialization")]` (https://serde.rs/container-attrs.html#into), implement `Into<FnDeclForSerialization>` for `FnDecl`, and call to getters like `is_async` or `is_generator` inside `into` implementation. In this way, `FnDecl` remains having getter methods instead of extra fields, while `FnDeclForSerialization` has extra fields in place of `FnDecl`. Unless we attempt to serialize it, memory usage can be kept micro-optimized.

I haven't tried the idea yet but I believe it works. What do you think about it?

Comment (dsherret) [reply rs-lib/src/generated.rs:6464]: @magurotuna yes, that would be perfect and allow for easily making the serialization a cargo feature.

I'm just still not sure about us creating an additional way of serializing swc's AST though so I wouldn't recommend going through the effort yet. I think you might be right that this should be done... have you run this by Bartek? It would be good to get his opinion on it if not.

Comment (magurotuna) [reply rs-lib/src/generated.rs:6464]: @dsherret 

> have you run this by Bartek? It would be good to get his opinion on it if not.

No, I haven't talked with Bartek about the implementation detail.
I heard Bartek is taking days off now, so let me ask his opinion later.

FWIW here is the current status: https://github.com/denoland/deno_lint/compare/main...magurotuna:plugin
[`Plugin` class in `plugin-base.ts`](https://github.com/denoland/deno_lint/compare/main...magurotuna:plugin#diff-049946972c1f48c2562e18fcfb007d7b25607cc56c44fc4e572523e29c28c880R171) is analogous to [`handler.rs` in Rust side](https://github.com/denoland/deno_lint/blob/main/src/handler.rs).
I know there exist swc's own TypeScript type definitions [here](https://github.com/swc-project/swc/blob/master/node-swc/src/types.ts), but they don't really seem useful for me as they don't exactly reflect the struct shape in Rust.
I have been thinking about how to improve deno_lint's plugin for about a week. At first I tried to leverage swc's `Serialize` in some way but I couldn't make it work - this is why I opened this PR honestly. 😌

Comment (dsherret) [reply rs-lib/src/generated.rs:6464]: @magurotuna sorry for my delay! I had a lot of work to do during the week.

I'm just looking right now and isn't doing an op for doing something like getting the parent or children going to be really expensive? I also think comparisons of nodes won't work.

Looking at the swc typescript nodes, it's too bad those don't have the parents at the current moment and that's something we could maybe ask for.

Maybe I'm completely misunderstanding this or overlooking something, but for deno lint's plugin design, what I would recommend is the following:

1. A single serialization of the AST to JSON.
2. This is passed to typescript code that has code generated typescript declarations based on the JSON structure as well as code generated helper methods like `#getChildren()`.
3. Once the JSON is parsed, the nodes are traversed and a `.parent` property is set for every node along with its prototype for the helper methods.

I think this would be a lot faster. I'm not sure how this works in the context of eventually having a type checker in Rust though. If we do it this way, would it be necessary for us to have this code in here and maybe instead we could work on improving swc's typescript code to have stuff like `.parent` node properties and helper methods? Thoughts?

That said, after thinking about it, I wouldn't be opposed to merging something for now that does this in this repo as long as it gets generated into a separate file and is hidden under a cargo flag. You may want to code generate stuff that does points 2 and 3 as well (if that's the case then maybe let's do it in a different PR. I would want to upgrade this library to use [code-block-writer](https://github.com/dsherret/code-block-writer) before that)... I think swc's current implementation is very node based instead of deno based.

Comment (magurotuna) [reply rs-lib/src/generated.rs:6464]: @dsherret 
Thanks again!
I'm not sure if I fully understand what you are suggesting, but that sounds generally good. Please correct me if I'm wrong - my understanding is that:
- serialization from AST (in Rust) fo JSON takes place just once, at the time when the plugin execution is invoked
- TypeScript side also has the logic for traversing which is similar to what `dprint-swc-ecma-ast-view` provides. We have `getChildren` methods implemented in TypeScript for each node type, like `ArrayLit`, `ForStmt`, etc.
- Also, each node has a `.parent` property whose value is acquired by traversing JSON after it is passed from Rust

It sounds really good and definitely has better performance than my design although we won't need to take plugin's performance into consideration too much in my opinion (at least in the early stage). Additionally, this way we could use swc's typescript code with only `.parent` and `getChildren` being added, so we could probably avoid unnecessary costs to maintain our types.
One thing I'm worried about is that it might be tough and challenging (at least for me) to port `dprint-swc-ecma-ast-view`'s logic to TypeScript. But I'm sure it's worth a try :)

> That said, after thinking about it, I wouldn't be opposed to merging something for now that does this in this repo as long as it gets generated into a separate file and is hidden under a cargo flag. You may want to code generate stuff that does points 2 and 3 as well (if that's the case then maybe let's do it in a different PR. I would want to upgrade this library to use code-block-writer before that)... I think swc's current implementation is very node based instead of deno based.

OK! Then I'll rewrite this PR to use `#[serde(into = "...")]`, let them generated into a separate file, and create a new cargo feature. Let's see if that will work fine and have enough performance.

Comment (dsherret) [reply rs-lib/src/generated.rs:6464]: @magurotuna yes. By the way, I wonder if the comments and tokens should be sent over? Maybe that part could be implemented last and we can not bother for now.

> One thing I'm worried about is that it might be tough and challenging (at least for me) to port dprint-swc-ecma-ast-view's logic to TypeScript. But I'm sure it's worth a try :)

Let me know if you would like some help! I can setup code-block-writer and some basic scaffolding once you get the PR in for Rust serialization and do some work on this as well.

> OK! Then I'll rewrite this PR to use #[serde(into = "...")], let them generated into a separate file, and create a new cargo feature. Let's see if that will work fine and have enough performance.

Sounds good! Thanks!

Comment (magurotuna) [reply rs-lib/src/generated.rs:6464]: @dsherret 

> By the way, I wonder if the comments and tokens should be sent over? Maybe that part could be implemented last and we can not bother for now.

Ideally they also should be sent over, but we'd have to come up with some way to serialize them into JSON well. So yeah, I agree as to implementing those things last.

> Let me know if you would like some help! I can setup code-block-writer and some basic scaffolding once you get the PR in for Rust serialization and do some work on this as well.

Your help will be really appreciated, thanks so much! :heart:

Comment (magurotuna) [reply rs-lib/src/generated.rs]: @dsherret 
Completed! Could you take a look again? (no rush 😉)

**Review by dsherret — APPROVED** [review]
@magurotuna this looks very good. Well done!

**Comment by dsherret** [comment]
Sorry for my delay on this! I had a busy week and then I forgot about it. I will do a release soon.

By the way, one point is I think it might be worthwhile to implement #14 in the future (let's not worry about it for now).
